src/utils.rs

```diff
@@ -87,34 +87,42 @@ pub fn bin2hex(hex: &mut [u8], bin: &[u8]) -> Result<(), ()> {
 /// and will produce the same output. Any hex characters part of `ignore` will not be ignored. Passing ": FC" in `ignore`
 /// when input is "69 : FC" will result in output "69FC".
 /// `bin_len` is the length of raw bytes that `hex` was converted to.
-/// `hex_end` is an optional parameter, if passed a pointer, it will be set to the address of the
+/// `hex_end` is an optional parameter, if passed it will be set to the index of the
 /// first byte after the last valid parsed character.
-/// `hex2bin()` will return Err<()> if `bin` is not sufficient to store the parsed string or if the
-/// string couldn't be fully parsed, but a valid pointer for `hex_end` was not provided
+/// `hex2bin()` will return Err<()> if `bin` is not sufficient to store the parsed string
 /// Refer docs at https://download.libsodium.org/doc/helpers#hexadecimal-encoding-decoding
 pub fn hex2bin(
     bin: &mut [u8],
     hex: &[u8],
     ignore: Option<&[u8]>,
     bin_len: &mut usize,
-    hex_end: Option<&mut [u8]>,
+    hex_end: Option<&mut isize>,
```

Comment:
Not good. hex_end should be a pointer to a pointer, so it’s wrong here.

Also exposing that is completely unidiomatic in a rust api, and idiomatic way would be to return ‘Result<&[u8]> where the return value is a sub slice of ‘hex’. But to be honest that’s a very C-ish way to write an api, and I don’t see the need for exposing ‘hex_end’ at all.